# Complete computational sequence characterization of mobile element variations in the human genome using meta-personal genome data


The developed tool is the first system that provides full sequence characterization of insertional structural variants in human genomes. It is capable of characterizing short mobile element insertions with a 92% accuracy. Overall, the characterization success rate for Alu is 75.03% with the mean of discordant and split-reads higher than 94 reads. For SVA, it is 71.43% with the threshold of 363 reads. And for L1 the values are 77.78% and 355 respectively.
The SV characterization depends on the allele frequency and is influenced by the repetitiveness of flanking regions. Therefore, addressing these problems is a key to further improvements.

Our method is based on collecting discordant, concordant and split-reads from all publicly available human genome data that cover breakpoints and an insertion. These reads are validated based on quality filters and patterns. Then, they are assembled into contigs using local de novo sequence assembly, and the contigs are merged from both sides of the insertion until they overlap. The information about a particular MEI is extracted from the alignment to the consensus, and TSD is deduced. In case of a successful characterization, the full insertion sequence is reported, otherwise, at least the breakpoint sequences are obtained.
